### PR TITLE
AVRO-1954 - Schema.Field.defaultVal() generates: Unknown datum type org.apache.avro.JsonProperties$Null

### PR DESCRIPTION
**Pull request by nandorKollar**
Added the test cases in the patch attached to AVRO-1954 Jira to TestGenericDatumWriter (since the issue is related to GenericDatumWriter), and modified GenericData to map JsonProperties.NULL_VALUE to Type.NULL

### CODE REVIEW

File: lang/java/avro/src/test/java/org/apache/avro/generic/TestGenericDatumWriter.java [hunk 1]
```diff
@@ -257,4 +256,79 @@ public void writeDoesNotAllowJavaEnumForGenericEnum() throws IOException {
...
+    Schema schema = Schema.createRecord("my_record", "doc", "mytest.namespace", false);
+    Schema.Field stringField = new Schema.Field("f",
+            Schema.createUnion(Schema.create(Schema.Type.NULL), Schema.create(Schema.Type.STRING)),
+            "field doc doc", org.apache.avro.JsonProperties.NULL_VALUE);
+    schema.setFields(Arrays.asList(stringField));
+    return schema;
+  }
+
+  private void writeObject(Schema schema, GenericRecord datum) throws Exception {
+    BinaryEncoder encoder = EncoderFactory.get().binaryEncoder(new ByteArrayOutputStream(5), null);
```
Comment (tomwhite) [inline lang/java/avro/src/test/java/org/apache/avro/generic/TestGenericDatumWriter.java]: Why 5? Just use the default unless there's a good reason not to.

File: lang/java/avro/src/test/java/org/apache/avro/generic/TestGenericDatumWriter.java [hunk 2]
```diff
@@ -257,4 +256,79 @@ public void writeDoesNotAllowJavaEnumForGenericEnum() throws IOException {
...
+
+  @Test
+  public void testExternalSchemaWithoutExplicitNullDefault() throws Exception {
+    Schema schema = createExternalSchemaWithoutExplicitNullDefault();
+    GenericRecord record = createRecord(schema);
+    writeObject(schema, record);
+  }
+
+  @Test
+  public void testCodeGeneratedSchemaWithoutExplicitNullDefault() throws Exception {
```
Comment (tomwhite) [inline lang/java/avro/src/test/java/org/apache/avro/generic/TestGenericDatumWriter.java]: I don't think both the testExternal and testCodeGenerated tests are needed as they only differ in how the schema is created.

File: lang/java/avro/src/test/java/org/apache/avro/generic/TestGenericDatumWriter.java [hunk 3]
```diff
@@ -257,4 +256,79 @@ public void writeDoesNotAllowJavaEnumForGenericEnum() throws IOException {
...
+    return schema;
+  }
+
+  private void writeObject(Schema schema, GenericRecord datum) throws Exception {
+    BinaryEncoder encoder = EncoderFactory.get().binaryEncoder(new ByteArrayOutputStream(5), null);
+    GenericDatumWriter<GenericData.Record> writter = new GenericDatumWriter<GenericData.Record>(schema);
+    writter.write(schema, datum, encoder);
+  }
+
+  private GenericRecord createRecord(Schema schema) {
```
Comment (tomwhite) [inline lang/java/avro/src/test/java/org/apache/avro/generic/TestGenericDatumWriter.java]: Call this `createRecordWithDefaultField` to make it clearer what it does.

File: lang/java/avro/src/test/java/org/apache/avro/generic/TestGenericDatumWriter.java (hunk 1 above)
Comment (nandorKollar) [inline lang/java/avro/src/test/java/org/apache/avro/generic/TestGenericDatumWriter.java]: Good point, I don't think there is a good reason for using 5. Removed the parameter, using the default now.

File: lang/java/avro/src/test/java/org/apache/avro/generic/TestGenericDatumWriter.java (hunk 3 above)
Comment (nandorKollar) [inline lang/java/avro/src/test/java/org/apache/avro/generic/TestGenericDatumWriter.java]: Renamed the method.

**Comment by derekstraka** [comment]
@nandorKollar - I ran into this issue today, and I'm interest in carrying this into master.  Any issues using your work as a starting point and adding you as signed off by on the commits?  I can run them by you first if you'd like.  Thanks

**Comment by nandorKollar** [comment]
@derekstraka I've no problem with using my work as base, but I'm afraid this pull request is already merged to master: https://github.com/apache/avro/blob/master/lang/java/avro/src/main/java/org/apache/avro/generic/GenericData.java#L730

I'm afraid it was left open here accidentally. Are you sure that you ran into the same issue?

**Comment by derekstraka** [comment]
Ah, awesome.  I was using the 1.8.0 release, and I had applied this patch to that tag locally and it resolved my issue.  I didn't think to check in master to see if it was already merged.  Thanks for your help!

**Comment by nandorKollar** [comment]
Great! However I think this should be backported to branch-1.8 too.

**Pull request closed without merging**